client/material/overlay.jsx

```diff
@@ -0,0 +1,112 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { CSSTransition } from 'react-transition-group'
+import styled from 'styled-components'
+
+import Popover from '../material/popover.jsx'
+import { fastOutSlowIn } from '../material/curve-constants.js'
+
+const transitionNames = {
+  appear: 'enter',
+  appearActive: 'enterActive',
+  enter: 'enter',
+  enterActive: 'enterActive',
+  exit: 'exit',
+  exitActive: 'exitActive',
+}
+
+const Contents = styled.div`
+  min-width: 256px;
+  position: relative;
+
+  .enter & {
+    opacity: 0;
+    transform: translateY(-16px);
+    transition-property: all;
+    /* TODO(tec27): this should be linear for opacity, and probably linearOutSlowIn for
+       translation? */
+    transition-timing-function: ${fastOutSlowIn};
+  }
+
+  .enterActive & {
+    opacity: 1;
+    transform: translateY(0px);
+    pointer-events: none;
+  }
+
+  .exit & {
+    opacity: 1;
+    transition-property: opacity;
+    transition-timing-function: linear;
+  }
+
+  .exitActive & {
+    opacity: 0;
+    pointer-events: none;
+  }
+`
+
+export default class Overlay extends React.Component {
+  static propTypes = {
+    open: PropTypes.bool.isRequired,
+    onDismiss: PropTypes.func.isRequired,
+    anchor: PropTypes.object,
+  }
+
+  render() {
+    const {
+      children,
+      open,
+      onDismiss,
+      anchor,
+      anchorOriginVertical,
+      anchorOriginHorizontal,
+      popoverOriginVertical,
+      popoverOriginHorizontal,
+    } = this.props
+
+    return (
+      <Popover
```

Comment:
When you have a component like this one that is only passing properties to the one that it is using, then you can just use the spread operator instead of specifying each prop individually, e.g.
```jsx
<Popover {...this.props} />
```
If you don't want to pass _all_ of the props to the underlying one, then you can do something like this:
```jsx
const { prop1, prop2, ...rest } = this.props

<Popover {...rest} />
```
So if a certain component doesn't use certain props, but just passes them down the chain, then using the spread operator makes it easier to update stuff later if a prop needs to be added/removed/renamed. Also, a component doesn't need to validate the prop types if it's not using those props, so you can remove that as well here.

Btw, this could also be applied in the above code too, like with passing the `onWhisperClick` handler down the chain. But I'm guilty of not doing this always myself :d, and I guess an argument could be made that it obfuscates things a bit.

Reply:
Now that I've actually seen this implemented, allow me to backtrack on this suggestion please ^^.  This is not the pattern that we should be using since it could very easily bite us in the ass (for example when new props are added, they could accidentally and unintentionally be passed down).

So there are two things I would change back here:
- revert the spreading of `onWhisperClick`. Just pass it down directly as it was before.
- make the `ProfileOverlay` component receive a prop called `popoverProps` explicitly (which is an object of all the Popover-specific properties) and then spread that instead, e.g in `profile-overlay.jsx`.
```jsx
const { children, popoverProps } = this.props

return (<Popover {...popoverProps}>...</Popover>)
```
Sorry about that :/

Reply:
No problem! I was actually wondering about that myself :P